schema/schemaspec/schemahcl/opts.go

```diff
@@ -37,10 +41,48 @@ func EvalContext(ctx *hcl.EvalContext) Option {
 }
 
 // WithTypes configures the list of given types as identifiers in the unmarshaling context.
-func WithTypes(types []*schemaspec.Type) Option {
+func WithTypes(typeSpecs []*schemaspec.TypeSpec) Option {
 	return func(config *Config) {
-		for _, t := range types {
-			config.ctx.Variables[t.Name] = cty.CapsuleVal(ctySchemaType, t)
+		for _, typeSpec := range typeSpecs {
+			if len(typeSpec.Attributes) == 0 {
+				typ := &schemaspec.Type{T: typeSpec.T}
+				config.ctx.Variables[typeSpec.Name] = cty.CapsuleVal(ctyTypeSpec, typ)
+				continue
+			}
+			spec := &function.Spec{
+				Type: function.StaticReturnType(ctyTypeSpec),
+			}
+			for _, arg := range typeSpec.Attributes {
+				p := function.Parameter{
+					Name:      arg.Name,
+					AllowNull: !arg.Required,
+				}
+				switch arg.Kind {
+				case reflect.String:
+					p.Type = cty.String
+				case reflect.Int, reflect.Float32:
+					p.Type = cty.Number
+				case reflect.Bool:
+					p.Type = cty.Bool
+				}
+				spec.Params = append(spec.Params, p)
+				spec.Impl = func(args []cty.Value, retType cty.Type) (cty.Value, error) {
+					t := &schemaspec.Type{
+						T: typeSpec.T,
+					}
+					for i, arg := range args {
+						a := schemaspec.Attr{K: typeSpec.Attributes[i].Name}
+						v, err := extractLiteralValue(arg)
+						if err != nil {
+							return cty.NilVal, err
+						}
+						a.V = v
+						t.Attributes = append(t.Attributes, &a)
```

Comment:
```suggestion
						v, err := extractLiteralValue(arg)
						if err != nil {
							return cty.NilVal, err
						}
						t.Attributes = append(t.Attributes, schemaspec.Attr{K: typeSpec.Attributes[i].Name, V: v})
```

It will be easier to read it this way.